clang/test/Driver/sycl-device-lib.cpp

```diff
@@ -122,3 +122,11 @@
 // SYCL_LLVM_LINK_NO_DEVICE_LIB: clang{{.*}} "-cc1" {{.*}} "-fsycl-is-device"
 // SYCL_LLVM_LINK_NO_DEVICE_LIB-NOT: llvm-link{{.*}}  "-only-needed"
 // SYCL_LLVM_LINK_NO_DEVICE_LIB: sycl-post-link{{.*}}  "-symbols" "-spec-const=rt" "-o" "{{.*}}.table" "{{.*}}.bc"
+
+/// ###########################################################################
+/// test llvm-link behavior for special user input whose filename resembles SYCL device library
+// RUN: touch libsycl-crt.o
+// RUN: %clangxx -fsycl libsycl-crt.o -### 2>&1 \
+// RUN:   | FileCheck %s -check-prefix=SYCL_LLVM_LINK_USER_ONLY_NEEDED
+// SYCL_LLVM_LINK_USER_ONLY_NEEDED: llvm-link{{.*}}  "{{.*}}.o" "-o" "{{.*}}.bc" "--suppress-warnings"
```

Comment:
Sorry for the confusion. Why is the name of the passed "user" object similar to a real device library, but it isn't recognized as a device library? Or is something else happening here that I'm misinterpreting?
```suggestion
// SYCL_LLVM_LINK_USER_ONLY_NEEDED: llvm-link{{.*}}  "libsycl-crt.o" "-o" "{{.*}}.bc" "--suppress-warnings"
```

Reply:
Hi, @AGindinson 
This is expected behavior and this is why crafting such a test is difficult. We input "libsycl-crt.o" to clang driver, clang driver will do following things:
1. use clang-offload-bundler tool to extract the internal LLVM IR from the user input and the extracted LLVM IR module's name will be libsycl-crt-XXXX.o where XXXX is a temp string
2. use llvm-link to link user's LLVM IR module which is the "step 1" link, the command used is like:
    llvm-link libsycl-crt-XXXX.o -o libsycl-crt-XXXX.bc
3. llvm-link to link user's LLVM IR module generated in 2 together with all SYCL device libraries, the input of llvm-link will be like:
    libsycl-crt-XXXX.bc(user's input) libsycl-crt-XXXX.o/obj(SYCL device libray) libsycl-cmath-XXXX.o/obj.....

As we can see, although user's input has the same name as "libsycl-crt.o", its postfix will be ".bc" instead of ".o" or ".obj", our check won't recognize it as SYCL device libraries.
Thanks very much.

Reply:
Thank you. I think I'm fully getting this now; apologies for suggesting a not-so-relevant test case then. It probably doesn't hurt, though.